### PR TITLE
Upgrade pending segments when a concurrent replace happens

### PR DESCRIPTION
**Pull request by kfaraz**
### Changes
- Add new task action to commit realtime segments when using APPEND locks
- Include `TaskLockType` in `SequenceMetadata`. The lock type is used to determine the right action type for committing realtime segments.
- Add utility method `IndexTaskUtils.emitSegmentPublishMetrics`
- Remove unused method `IndexTaskUtils.getMessagesFromSavedParseExceptions`

### Pending changes
- Add method in `IndexerMetadataStorageCoordinator` to commit segments, commit metadata and upgrade segments in the same transaction
- Fix `commitReplaceSegments` method to identify pending segments to upgrade and create upgraded entries

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/common/actions/CommitRealtimeSegmentsAndMetadataAction.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+ * This action performs the following operations in a single transaction:
+ * <ul>
+ * <li>Commit the append segments</li>
+ * <li>Upgrade the append segments to all visible REPLACE versions</li>
+ * <li>Commit start and end {@link DataSourceMetadata}.</li>
+ * </ul>
+ * This action differs from {@link SegmentTransactionalInsertAction} as it is used
+ * only with APPEND locks and also upgrades segments as needed.
+ */
+public class CommitRealtimeSegmentsAndMetadataAction implements TaskAction<SegmentPublishResult>
```
Comment (AmatyaAvadhanula) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/actions/CommitRealtimeSegmentsAndMetadataAction.java:55]: Could we not re-use SegmentTransactionAppendAction with the metadata being null for batch and the required values for streaming ingestion (similar to the original insert action)?

Is there anything besides the metadata commit that this action does that the transactional append action doesn't?

Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/actions/CommitRealtimeSegmentsAndMetadataAction.java:55]: Yeah, I was thinking the same. Let me see what we can do.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SequenceMetadata.java [hunk 2]
```diff
@@ -73,7 +78,8 @@ public SequenceMetadata(
       @JsonProperty("startOffsets") Map<PartitionIdType, SequenceOffsetType> startOffsets,
       @JsonProperty("endOffsets") Map<PartitionIdType, SequenceOffsetType> endOffsets,
       @JsonProperty("checkpointed") boolean checkpointed,
-      @JsonProperty("exclusiveStartPartitions") Set<PartitionIdType> exclusiveStartPartitions
+      @JsonProperty("exclusiveStartPartitions") Set<PartitionIdType> exclusiveStartPartitions,
+      @JsonProperty("taskLockType") TaskLockType taskLockType
```
Comment (abhishekagarwal87) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SequenceMetadata.java:82]: This should be nullable for backward compatibility reasons. should it not?

File: server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java [hunk 3]
```diff
@@ -971,6 +966,70 @@ private static class CheckExistingSegmentIdResult
...
+  )
+  {
+    verifySegmentsToCommit(appendSegments);
+    if ((startMetadata == null && endMetadata != null)
+        || (startMetadata != null && endMetadata == null)) {
+      throw new IllegalArgumentException("start/end metadata pair must be either null or non-null");
+    }
+
+    final String dataSource = appendSegments.iterator().next().getDataSource();
+    final Set<DataSegment> upgradedSegments = connector.retryTransaction(
```
Comment (abhishekagarwal87) [inline server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:983]:
```suggestion
    final Set<DataSegment> segmentsToUpgrade = connector.retryTransaction(
```

**Comment by kfaraz** [comment]
Closing this PR as these changes have already been merged as a part of #15039

**Pull request closed without merging**